### ISSUE TITLE
Assertion `options.use_os_buffer' failed.

### ISSUE THREAD

**Issue by RoelVdP** [facebook/mysql-5.6#221]
```
Version: '5.6.27-debug'  socket: '/sda/FB-RocksDB/socket.sock'  port: 11526  MySQL Community Server (GPL)
2016-03-24 09:58:20 26936 [Note] RocksDB: Begin filtering dropped index (0,256)
mysqld: /git/FB-mysql-5.6/rocksdb/util/io_posix.cc:238: rocksdb::PosixMmapReadableFile::PosixMmapReadableFile(int, const string&, void*, size_t, const rocksdb::EnvOptions&): Assertion `options.use_os_buffer' failed.
22:58:20 UTC - mysqld got signal 6 ;
```

```
Program terminated with signal 6, Aborted.
#0  0x00007fc1592b0741 in __pthread_kill (threadid=<optimized out>, signo=6) at ../nptl/sysdeps/unix/sysv/linux/pthread_kill.c:61
61        val = INTERNAL_SYSCALL (tgkill, err, 3, THREAD_GETMEM (THREAD_SELF, pid),
(gdb) bt
#0  0x00007fc1592b0741 in __pthread_kill (threadid=<optimized out>, signo=6) at ../nptl/sysdeps/unix/sysv/linux/pthread_kill.c:61
#1  0x0000000000e8c11a in my_write_core (sig=6) at /git/FB-mysql-5.6/mysys/stacktrace.c:422
#2  0x0000000000b01ae5 in handle_fatal_signal (sig=6) at /git/FB-mysql-5.6/sql/signal_handler.cc:239
#3  <signal handler called>
#4  0x00007fc15723d5f7 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#5  0x00007fc15723ece8 in __GI_abort () at abort.c:90
#6  0x00007fc157236566 in __assert_fail_base (fmt=0x7fc157386228 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x174049f "options.use_os_buffer", file=file@entry=0x17403f0 "/git/FB-mysql-5.6/rocksdb/util/io_posix.cc", line=line@entry=238, function=function@entry=0x1741340 <rocksdb::PosixMmapReadableFile::PosixMmapReadableFile(int, std::string const&, void*, unsigned long, rocksdb::EnvOptions const&)::__PRETTY_FUNCTION__> "rocksdb::PosixMmapReadableFile::PosixMmapReadableFile(int, const string&, void*, size_t, const rocksdb::EnvOptions&)") at assert.c:92
#7  0x00007fc157236612 in __GI___assert_fail (assertion=0x174049f "options.use_os_buffer", file=0x17403f0 "/git/FB-mysql-5.6/rocksdb/util/io_posix.cc", line=238, function=0x1741340 <rocksdb::PosixMmapReadableFile::PosixMmapReadableFile(int, std::string const&, void*, unsigned long, rocksdb::EnvOptions const&)::__PRETTY_FUNCTION__> "rocksdb::PosixMmapReadableFile::PosixMmapReadableFile(int, const string&, void*, size_t, const rocksdb::EnvOptions&)") at assert.c:101
#8  0x0000000001054a29 in rocksdb::PosixMmapReadableFile::PosixMmapReadableFile (this=0x7fc1502150f0, fd=32, fname="./.rocksdb/000012.sst", base=0x7fc159929000, length=708, options=...) at /git/FB-mysql-5.6/rocksdb/util/io_posix.cc:238
#9  0x000000000104c47f in rocksdb::(anonymous namespace)::PosixEnv::NewRandomAccessFile (this=0x2018740 <rocksdb::Env::Default()::default_env>, fname="./.rocksdb/000012.sst", result=0x7fc150dfbfa0, options=...) at /git/FB-mysql-5.6/rocksdb/util/env_posix.cc:186
#10 0x0000000000fb7e5e in rocksdb::TableCache::GetTableReader (this=0x7fc1552cee00, env_options=..., internal_comparator=..., fd=..., sequential_mode=false, record_read_stats=true, file_read_hist=0x7fc15532f000, table_reader=0x7fc150dfc0c0, skip_filters=false) at /git/FB-mysql-5.6/rocksdb/db/table_cache.cc:95
#11 0x0000000000fb83f7 in rocksdb::TableCache::FindTable (this=0x7fc1552cee00, env_options=..., internal_comparator=..., fd=..., handle=0x7fc150dfc220, no_io=false, record_read_stats=true, file_read_hist=0x7fc15532f000, skip_filters=false) at /git/FB-mysql-5.6/rocksdb/db/table_cache.cc:137
#12 0x0000000000fb8816 in rocksdb::TableCache::NewIterator (this=0x7fc1552cee00, options=..., env_options=..., icomparator=..., fd=..., table_reader_ptr=0x0, file_read_hist=0x7fc15532f000, for_compaction=false, arena=0x0, skip_filters=false) at /git/FB-mysql-5.6/rocksdb/db/table_cache.cc:187
#13 0x00000000010cd4e0 in rocksdb::BuildTable (dbname="./.rocksdb", env=0x2018740 <rocksdb::Env::Default()::default_env>, ioptions=..., env_options=..., table_cache=0x7fc1552cee00, iter=0x7fc150dfc9e8, meta=0x7fc150dfd4d0, internal_comparator=..., int_tbl_prop_collector_factories=0x7fc15530fe40, column_family_id=0, snapshots=std::vector of length 0, capacity 0, earliest_write_conflict_snapshot=72057594037927935, compression=rocksdb::kNoCompression, compression_opts=..., paranoid_file_checks=false, internal_stats=0x7fc153bbda00, io_priority=rocksdb::Env::IO_HIGH, table_properties=0x7fc150dfd818) at /git/FB-mysql-5.6/rocksdb/db/builder.cc:152
#14 0x0000000000f8ee24 in rocksdb::FlushJob::WriteLevel0Table (this=0x7fc150dfd780, mems=..., edit=0x7fc155246200, meta=0x7fc150dfd4d0) at /git/FB-mysql-5.6/rocksdb/db/flush_job.cc:237
#15 0x0000000000f8e1c3 in rocksdb::FlushJob::Run (this=0x7fc150dfd780, file_meta=0x7fc150dfd700) at /git/FB-mysql-5.6/rocksdb/db/flush_job.cc:148
#16 0x0000000000f26545 in rocksdb::DBImpl::FlushMemTableToOutputFile (this=0x7fc155382e00, cfd=0x7fc15530fe00, mutable_cf_options=..., made_progress=0x7fc150dfeabf, job_context=0x7fc150dfdfe0, log_buffer=0x7fc150dfe1b0) at /git/FB-mysql-5.6/rocksdb/db/db_impl.cc:1498
#17 0x0000000000f2c6ae in rocksdb::DBImpl::BackgroundFlush (this=0x7fc155382e00, made_progress=0x7fc150dfeabf, job_context=0x7fc150dfdfe0, log_buffer=0x7fc150dfe1b0) at /git/FB-mysql-5.6/rocksdb/db/db_impl.cc:2656
#18 0x0000000000f2c869 in rocksdb::DBImpl::BackgroundCallFlush (this=0x7fc155382e00) at /git/FB-mysql-5.6/rocksdb/db/db_impl.cc:2678
#19 0x0000000000f2c1c4 in rocksdb::DBImpl::BGWorkFlush (db=0x7fc155382e00) at /git/FB-mysql-5.6/rocksdb/db/db_impl.cc:2594
#20 0x00000000010578eb in rocksdb::ThreadPool::BGThread (this=0x7fc15522b2a0, thread_id=0) at /git/FB-mysql-5.6/rocksdb/util/thread_posix.cc:119
#21 0x0000000001057968 in rocksdb::BGThreadWrapper (arg=0x7fc1552150a0) at /git/FB-mysql-5.6/rocksdb/util/thread_posix.cc:143
#22 0x00007fc1592abdc5 in start_thread (arg=0x7fc150dff700) at pthread_create.c:308
#23 0x00007fc1572fe21d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:113
```

```
Options required (note --rocksdb-allow-os-buffer=0 and --rocksdb-allow-mmap-reads=1):
--no-defaults --default-tmp-storage-engine=MyISAM --rocksdb --skip-innodb --default-storage-engine=RocksDB --rocksdb-allow-os-buffer=0 --rocksdb-allow-mmap-reads=1
```

```
Testcase:
DROP DATABASE test;CREATE DATABASE test;USE test;
create TABLE t1(p int,a char collate latin1_swedish_ci)ENGINE=RocksDB;
INSERT INTO t1 VALUES(NOW(),1);
drop table if exists t1,v1;
SELECT 1;   # Shows server is gone
```

**Comment by jkedgar** [comment]
According to RocksDB engineers, allow_mmap_reads implies allow_os_buffer (which defaults to on) since by using memory mapped files you are automatically allowing the OS to cache the data, so intentionally turning it off is not valid.  I've added a diff for RocksDB that will cause it to fail to open the database if this combination of options is used (https://reviews.facebook.net/D56397).  Then I added a diff for MyRocks to also validate and not start the server if this combination of options is used (https://reviews.facebook.net/D56403).  As you are working with MyRocks you only need the second one.
